File: ObasAnalyzerVB/ObasAnalyzerVB/ObasAnalyzerVB/RepositorioGrandeAnalyzer.vb
Imports ObasAnalyzerVB.Utilerias.ObasAnalyzer

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class RepositorioGrandeAnalyzer
    Inherits DiagnosticAnalyzer

#Region "Reglas"
    Friend Const Category As String = "OBAS-SmellArquitectura"

    Friend Shared ReadOnly Regla001RepositorioGrande As DiagnosticDescriptor = New DiagnosticDescriptor(
            "RA11001",
            "RA11-001: Repositorio gestiona muchas entidades ",
            "El repositorio {0} gestiona {1} entidades y sobre pasa el límite definido para la cantidad de entidades que puede gestionar.",
            Category,
            DiagnosticSeverity.Error,
            True)


    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Regla001RepositorioGrande)
        End Get
    End Property

    Public Overrides Sub Initialize(ByVal context As AnalysisContext)
        ' Registra el tipo de sintaxis que se analiza para cada regla
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeRegla001RepositorioGrande, SyntaxKind.ClassBlock)
    End Sub
#End Region

#Region "Analyzadores"
    ''' <summary>
    ''' Valida el cumplimiento de la regla
    ''' "RA11-001: Repositorio gestiona muchas entidades"
    ''' </summary>
    ''' <param name="context"></param>
    Private Sub AnalyzeRegla001RepositorioGrande(ByVal context As SyntaxNodeAnalysisContext)
        ' Objeto para obtener entidades
        Dim entidades As New HashSet(Of TypeSyntax)()

        ' Obtiene la declaración de la clase
        Dim classDeclaration = DirectCast(context.Node, ClassBlockSyntax)

        ' Obtiene el símbolo de la clase
        Dim classSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration)

        ' Obtiene nombre de la clase
        Dim nombreClase = classDeclaration.BlockStatement.Identifier.ValueText

        ' Revisa si el nombre de la clase contenedora inicia con la cadena "dal"
        If nombreClase.ToLower().Contains(Constantes.nomenclaturaAccesoBaseDatos) Then
            ' Obtiene la lista de métodos
            Dim metodos = classDeclaration.Members.OfType(Of MethodBlockSyntax)()

            ' Revisa cada método
            For Each metodo In metodos
                ' Obtiene el cuerpo del método
                Dim cuerpoMetodo = metodo.Statements

                If cuerpoMetodo.Count > 0 Then
                    AnalizaCuerpo(cuerpoMetodo, entidades, context)
                End If
            Next

            ' Si sobrepasa el límite, reporte el error
            If entidades.Count > Constantes.limiteEntidadesRepositorio Then
                Dim diag As Diagnostic = Diagnostic.Create(Regla001RepositorioGrande, classDeclaration.BlockStatement.Identifier.GetLocation(), classSymbol.Name, entidades.Count)
                context.ReportDiagnostic(diag)
            End If
        End If
    End Sub

    ''' <summary>
    ''' Metodo que extrae los identificadores de los nodos del cuerpo de un metodo y luego
    ''' compara si el nombre inicia con Ent para determinar que se refiere a una entidad
    ''' </summary>
    ''' <param name="cuerpo"></param>
    '''  <param name="entidades"></param>
    ''' <param name="context"></param>
    Private Shared Sub AnalizaCuerpo(cuerpo As SyntaxList(Of StatementSyntax), entidades As HashSet(Of TypeSyntax), context As SyntaxNodeAnalysisContext)
        ' Revisa cada statement
        For Each statement As StatementSyntax In cuerpo
            ' Si es un nodo de expresión literal
            If TypeOf statement Is LocalDeclarationStatementSyntax Then
                Dim tipoVariable = DirectCast(statement, LocalDeclarationStatementSyntax).Declarators.FirstOrDefault().AsClause.Type

                If (tipoVariable IsNot Nothing) AndAlso (tipoVariable.ToString.ToLower().StartsWith(Constantes.nomenclaturaEntidad)) Then
                    If Not entidades.Contains(tipoVariable) Then
                        entidades.Add(tipoVariable)
                    End If
                End If
            End If
        Next
    End Sub



#End Region

End Class

File: ObasAnalyzerCSharp/ObasAnalyzerCSharp/ObasAnalyzerCSharp/RepositorioGrandeAnalyzer.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using Utilerias.ObasAnalyzerCSharp;

namespace ObasAnalyzerCSharp
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class RepositorioGrandeAnalyzer : DiagnosticAnalyzer
    {
        #region "Reglas"
        internal const string Category = "OBAS-SmellArquitectura-MVC";

        internal static readonly DiagnosticDescriptor Regla001RepositorioGrande = new DiagnosticDescriptor(
            "RA11001",
            "RA11-001: Repositorio gestiona muchas entidades ",
            "El repositorio {0} gestiona {1} entidades y sobre pasa el límite definido para la cantidad de entidades que puede gestionar.",
            Category,
            DiagnosticSeverity.Error,
            true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Regla001RepositorioGrande); } }

        public override void Initialize(AnalysisContext context)
        {
            context.EnableConcurrentExecution();
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
            // Registra el tipo de sintaxis que se analiza para cada regla
            context.RegisterSyntaxNodeAction(AnalyzeRegla001RepositorioGrande, SyntaxKind.ClassDeclaration);
        }
        #endregion

        #region "Analizadores"

        /// <summary>
        /// Valida el cumplimiento de la regla
        /// "RA11-001: Repositorio gestiona muchas entidades"
        /// </summary>
        /// <param name="context"></param>
        private void AnalyzeRegla001RepositorioGrande(SyntaxNodeAnalysisContext context)
        {
            // Objeto para obtener entidades
            var entidades = new HashSet<INamedTypeSymbol>();

            // Obtiene la declaración de la clase
            var classDeclaration = (ClassDeclarationSyntax)context.Node;

            // Obtiene el simbolo de la clase
            var classSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration);

            // Obtiene nombre de la clase
            var nombreClase = classDeclaration.Identifier.ValueText;

            // Revisa si el nombre de la clase contenedora inicia con la cadena "dal"
            if (nombreClase.ToLower().Contains(Constantes.nomenclaturaRepositorio))
            {
                // Obtiene la lista de metodos
                var metodos = classDeclaration.Members.OfType<BaseMethodDeclarationSyntax>();

                // Revisa cada metodo
                foreach (var metodo in metodos)
                {
                    // Obtiene el cuerpo del método
                    var cuerpoMetodo = metodo.Body;

                    if (cuerpoMetodo != null)
                    {
                        AnalizaCuerpo(cuerpoMetodo, entidades, context);
                    }
                }

                // Si sobrepasa el limite reporte el error
                if (entidades.Count > Constantes.limiteEntidadesRepositorio)
                {
                    var diagnostic = Diagnostic.Create(Regla001RepositorioGrande, classDeclaration.Identifier.GetLocation(), classSymbol.Name, entidades.Count);
                    context.ReportDiagnostic(diagnostic);
                }
            }
        }

        /// <summary>
        /// Metodo que extrae los identificadores de los nodos del cuerpo de un metodo y luego
        /// compara si el nombre inicia con Ent para determinar que se refiere a una entidad
        /// </summary>
        /// <param name="cuerpo"></param>
        /// <param name="entidades"></param>
        /// <param name="context"></param>
        private static void AnalizaCuerpo(CSharpSyntaxNode cuerpo, HashSet<INamedTypeSymbol> entidades, SyntaxNodeAnalysisContext context)
        {
            // Obtiene identificadores de los nodos
            var identificadores = cuerpo.DescendantNodes().OfType<IdentifierNameSyntax>();

            // Revisa los identificadores
            foreach (var identificador in identificadores)
            {
                var simboloTipo = context.SemanticModel.GetSymbolInfo(identificador).Symbol as INamedTypeSymbol;

                if (simboloTipo != null)
                {
                    if (simboloTipo != null && simboloTipo.Name.ToLower().StartsWith(Constantes.nomenclaturaEntidad))
                    {
                        // Añade la entidad al conteo
                        entidades.Add(simboloTipo);
                    }
                }
            }
        }
    }

    #endregion
}